<a href="https://colab.research.google.com/github/Wolf123987/Prediction-of-TCS-share-price-using-LSTM-method-and-obtain-share-price-data-from-yahoo-finance/blob/main/To_get_live_shares_price_of_TCS_and_predict_its_closing_price_by_LSTM_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
%%html
<marquee style='width: 60%;color: red;'><b>To get latest stock price  using python </b></marquee>

In [ ]:
pip install yfinance

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 62 kB 1.4 MB/s 
  Attempting uninstall: requests
    Found existing installation: requests 2.23.0
    Uninstalling requests-2.23.0:
      Successfully uninstalled requests-2.23.0


In [ ]:
import pandas as pd
import yfinance as yf
import datetime
from datetime import date, timedelta
today = date.today()



To obatail current price of TCS : we took its symbol ***TCS.NS*** means NSE data of TCS 


In [ ]:
d1 = today.strftime("%Y-%m-%d")
end_date = d1
d2 = date.today() - timedelta(days=100)
d2 = d2.strftime("%Y-%m-%d")
start_date = d2

manoj = yf.download('TCS.NS', 
                      start=start_date, 
                      end=end_date, 
                      progress=False)
print(manoj.tail(50))

                                  Open         High          Low        Close  \
Date                                                                            
2022-08-03 00:00:00+05:30  3280.000000  3349.000000  3277.050049  3339.500000   
2022-08-04 00:00:00+05:30  3350.000000  3387.899902  3307.350098  3354.949951   
2022-08-05 00:00:00+05:30  3361.199951  3377.000000  3355.000000  3365.050049   
2022-08-08 00:00:00+05:30  3365.000000  3378.899902  3336.000000  3374.449951   
2022-08-10 00:00:00+05:30  3385.000000  3385.000000  3335.000000  3354.250000   
2022-08-11 00:00:00+05:30  3400.000000  3428.699951  3375.050049  3422.500000   
2022-08-12 00:00:00+05:30  3419.000000  3419.000000  3381.250000  3401.550049   
2022-08-16 00:00:00+05:30  3411.000000  3414.850098  3387.500000  3392.699951   
2022-08-17 00:00:00+05:30  3385.100098  3417.949951  3371.100098  3401.100098   
2022-08-18 00:00:00+05:30  3390.000000  3392.000000  3362.000000  3381.250000   
2022-08-19 00:00:00+05:30  3

In [ ]:
manoj["Date"] = manoj.index
manoj = manoj[["Date", "Open", "High", 
             "Low", "Close", "Adj Close", "Volume"]]
manoj.reset_index(drop=True, inplace=True)
print(manoj.head())

                       Date         Open         High      Low        Close  \
0 2022-07-11 00:00:00+05:30  3206.149902  3225.000000  3106.00  3113.800049   
1 2022-07-12 00:00:00+05:30  3114.899902  3136.199951  3080.25  3084.699951   
2 2022-07-13 00:00:00+05:30  3104.000000  3110.000000  3035.00  3038.750000   
3 2022-07-14 00:00:00+05:30  3056.000000  3057.000000  2967.00  2998.750000   
4 2022-07-15 00:00:00+05:30  3018.550049  3028.899902  2953.00  2994.600098   

     Adj Close   Volume  
0  3097.585938  6974600  
1  3068.637207  3734815  
2  3022.926514  3863530  
3  2991.009033  4764908  
4  2986.869873  4574806  


**plot of Date Vs Closing price of stock**

In [ ]:
import plotly.express as px
import plotly.graph_objects as go
figure = px.scatter(data_frame = manoj, x="Date",
                    y="Close" ,trendline="ols")
figure.show()

***plot of date Vs hight price of stock ***

In [ ]:
import plotly.express as px
import plotly.graph_objects as go
figure = px.scatter(data_frame = manoj, x="Date",
                    y="High" ,trendline="ols")
figure.show()

here is the link to get  symbols of some important stocks of national stock exchange from yahooo finance website  just repale the stock symbol like here in this program i  have taken TCS stock 
https://finance.yahoo.com/quote/%5ENSEI/components/



***Lets see the candlestick chart of TCS stock ***

In [ ]:
import plotly.graph_objects as go
figure = go.Figure(data=[go.Candlestick(x=manoj["Date"],
                                        open=manoj["Open"], 
                                        high=manoj["High"],
                                        low=manoj["Low"], 
                                        close=manoj["Close"])])
figure.update_layout(title = "TCS  Stock Price Analysis", 
                     xaxis_rangeslider_visible=False)
figure.show()

In [ ]:
correlation = manoj.corr()
print(correlation["Close"].sort_values(ascending=False))

Close        1.000000
Adj Close    0.999885
High         0.985750
Low          0.984472
Open         0.957477
Volume      -0.469572
Name: Close, dtype: float64


Lets try to predict the closing price of TCS by LSTM( long short-term  memory)

In [ ]:
x = manoj[["Open", "High", "Low", "Volume"]]
y = manoj["Close"]
x = x.to_numpy()
y = y.to_numpy()
y = y.reshape(-1, 1)

In [ ]:
from sklearn.model_selection import train_test_split
xtrain, xtest, ytrain, ytest = train_test_split(x, y, test_size=0.2, random_state=42)

neural network architecture for LSTM

In [ ]:
from keras.models import Sequential
from keras.layers import Dense, LSTM

In [ ]:
model = Sequential()
model.add(LSTM(128, return_sequences=True, input_shape= (xtrain.shape[1], 1)))
model.add(LSTM(64, return_sequences=False))
model.add(Dense(25))
model.add(Dense(1))
model.summary()

Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 lstm_4 (LSTM)               (None, 4, 128)            66560     
                                                                 
 lstm_5 (LSTM)               (None, 64)                49408     
                                                                 
 dense_4 (Dense)             (None, 25)                1625      
                                                                 
 dense_5 (Dense)             (None, 1)                 26        
                                                                 
Total params: 117,619
Trainable params: 117,619
Non-trainable params: 0
_________________________________________________________________


In [ ]:
model.compile(optimizer='adam', loss='mean_squared_error')
model.fit(xtrain, ytrain, batch_size=1, epochs=30)

Epoch 1/30
53/53 [==============================] - 3s 5ms/step - loss: 9944778.0000
Epoch 2/30
53/53 [==============================] - 0s 5ms/step - loss: 9660500.0000
Epoch 3/30
53/53 [==============================] - 0s 5ms/step - loss: 9277011.0000
Epoch 4/30
53/53 [==============================] - 0s 5ms/step - loss: 8784380.0000
Epoch 5/30
53/53 [==============================] - 0s 5ms/step - loss: 8193021.5000
Epoch 6/30
53/53 [==============================] - 0s 5ms/step - loss: 7517844.0000
Epoch 7/30
53/53 [==============================] - 0s 5ms/step - loss: 6780827.0000
Epoch 8/30
53/53 [==============================] - 0s 5ms/step - loss: 6005355.5000
Epoch 9/30
53/53 [==============================] - 0s 5ms/step - loss: 5217704.0000
Epoch 10/30
53/53 [==============================] - 0s 5ms/step - loss: 4440636.5000
Epoch 11/30
53/53 [==============================] - 0s 5ms/step - loss: 3698359.0000
Epoch 12/30
53/53 [==============================] - 0s 5ms/ste

In [ ]:
import numpy as np
#features = [Open, High, Low, Adj Close, Volume]
features = np.array([[3100.0896, 3190.4198, 3095.0707,3120.04578, 3834845]])
model.predict(features)

1/1 [==============================] - 0s 25ms/step


array([[3151.6848]], dtype=float32)

# The ***predicted closing price of TCs is 3151.6848*** for a opening price of 3100.0896, day high of 3190.4198,day low of 3095.0707, previous day close of 3120.04578 and traded vloume of 3834845 shares 